### PR TITLE
Avoid KeyNotFound when updating inline rename state

### LINKED ISSUE #31883: crash during code analysis -- RenameTracking
<p>crash during automated test run.  crashed after around 8 hours.</p>

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/405103/crash-during-code-analysis-rename.html
VSTS ticketId: 748636_
_These are the original issue comments:_

sean on 12/11/2018, 09:30 AM (6 days ago): <p><a target='_blank' href="https://developercommunity.visualstudio.com/storage/temp/51819-devenvexe19864.zip">devenvexe19864.zip</a></p>

sean on 12/12/2018, 11:42 AM (5 days ago): <p>crashed last night also -- in Editor RenameTracking while a modal dialog is being displayed.</p><p>callstack attached as text file because the reporting system won't allow it to be pasted into this window due to it being greater than 10000 characters: <a target='_blank' href="https://developercommunity.visualstudio.com/storage/temp/52071-callstack.txt">callstack.txt</a></p><p>full heap dump is also attached: <a target='_blank' href="https://developercommunity.visualstudio.com/storage/temp/52072-v16p-w10-20181212-1325dmp.zip">v16p-w10-20181212-1325dmp.zip</a></p>

sean on 12/17/2018, 01:16 PM (4 hours ago): <p>Another crash during RenameTracking</p><p><a target='_blank' href="https://developercommunity.visualstudio.com/storage/temp/52765-v16p-w10-20181215-1257.zip">v16p-w10-20181215-1257.zip</a></p>

_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by JoeRobich**
Check to ensure our subject buffer has undo state before trying to update.

Fixes #31883

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.StateMachine.cs [hunk 1]
```diff
@@ -200,9 +200,8 @@ public bool ClearTrackingSession()
                 {
                     // Disallow the existing TrackingSession from triggering IdentifierFound.
                     var previousTrackingSession = this.TrackingSession;
-                    this.TrackingSession = null;
-
                     previousTrackingSession.Cancel();
+                    this.TrackingSession = null;
```
[the remark this answers is not in the document]
Comment (JoeRobich) [reply src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.StateMachine.cs]: It is. Should I break this up?

**Comment by sharwell** [comment]
❓ @JoeRobich Are you sure this isn't just a duplicate of #30661?

File: src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.StateMachine.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (genlu) [reply src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.StateMachine.cs]: No need. But if this is not a dupe of #30661 as @sharwell suggested, we probably need a comment here.

**Comment by JoeRobich** [comment]
@genlu Thanks.

@sharwell I agree the NRE is the same issue. I will remove those changes from this PR.

File: src/VisualStudio/Core/Def/Implementation/InlineRename/InlineRenameUndoManager.cs [hunk 2]
```diff
@@ -179,7 +179,15 @@ protected override void UndoTemporaryEdits(ITextBuffer subjectBuffer, bool disco
...
+                // See bug https://github.com/dotnet/roslyn/issues/31883.
+                if (!this.UndoManagers.TryGetValue(subjectBuffer, out var bufferUndoState))
+                {
+                    return;
+                }
+
+                ApplyReplacementText(subjectBuffer, bufferUndoState.TextUndoHistory, propagateSpansEditTag, spans, this.currentState.ReplacementText);
 
                 // Here we create the descriptions for the redo list dropdown.
                 var undoManager = this.UndoManagers[subjectBuffer].UndoManager;
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/InlineRename/InlineRenameUndoManager.cs]: 💡 Can we use the value we obtained above?

```suggestion
                 var undoManager = bufferUndoState.UndoManager;
```

**Pull request merged**